File: src/TypescriptSDK/src/models/ProjectInfo.ts
import { IsString, IsOptional, Matches, IsNumber, Min, Max, IsArray, IsInstance, ValidateNested, IsEnum, validate, ValidationError as TsValidationError } from 'class-validator';
import { Type, instanceToPlain, Expose, Transform } from 'class-transformer';
import { deepTransform } from '../deepTransform';
import { _OpenAPIGenBaseModel } from "./_OpenAPIGenBaseModel";
import { BuildingTypes } from "./BuildingTypes";
import { ClimateZones } from "./ClimateZones";
import { EfficiencyStandards } from "./EfficiencyStandards";
import { Location } from "./Location";

/** Project information. */
export class ProjectInfo extends _OpenAPIGenBaseModel {
    @Type(() => String)
    @IsString()
    @IsOptional()
    @Matches(/^ProjectInfo$/)
    @Expose({ name: "type" })
    /** type */
    type: string = "ProjectInfo";
	
    @Type(() => Number)
    @IsNumber()
    @IsOptional()
    @Min(-360)
    @Max(360)
    @Expose({ name: "north" })
    /** A number between -360 to 360 where positive values rotate the compass counterclockwise (towards the West) and negative values rotate the compass clockwise (towards the East). */
    north: number = 0;
	
    @IsArray()
    @Type(() => String)
    @IsString({ each: true })
    @IsOptional()
    @Expose({ name: "weather_urls" })
    /** A list of URLs to zip files that includes EPW, DDY and STAT files. You can find these URLs from the EPWMAP. The first URL will be used as the primary weather file. */
    weatherUrls?: string[];
	
    @Type(() => Location)
    @IsInstance(Location)
    @ValidateNested()
    @IsOptional()
    @Expose({ name: "location" })
    /** Project location. This value is usually generated from the information in the weather files. */
    location?: Location;
	
    @Type(() => String)
    @IsEnum(ClimateZones)
    @IsOptional()
    @Expose({ name: "ashrae_climate_zone" })
    /** Project location climate zone. */
    ashraeClimateZone?: ClimateZones;
	
    @IsArray()
    @Type(() => String)
    @IsEnum(BuildingTypes, { each: true })
    @IsOptional()
    @Expose({ name: "building_type" })
    /** A list of building types for the project. The first building type is considered the primary type for the project. */
    buildingType?: BuildingTypes[];
	
    @IsArray()
    @Type(() => String)
    @IsEnum(EfficiencyStandards, { each: true })
    @IsOptional()
    @Expose({ name: "vintage" })
    /** A list of building vintages (e.g. ASHRAE_2019, ASHRAE_2016). */
    vintage?: EfficiencyStandards[];
	

    constructor() {
        super();
        this.type = "ProjectInfo";
        this.north = 0;
    }


    override init(_data?: any) {

        if (_data) {
            const obj = deepTransform(ProjectInfo, _data);
            this.type = obj.type ?? "ProjectInfo";
            this.north = obj.north ?? 0;
            this.weatherUrls = obj.weatherUrls;
            this.location = obj.location;
            this.ashraeClimateZone = obj.ashraeClimateZone;
            this.buildingType = obj.buildingType;
            this.vintage = obj.vintage;
        }
    }


    static override fromJS(data: any): ProjectInfo {
        data = typeof data === 'object' ? data : {};

        if (Array.isArray(data)) {
            const obj:any = {};
            for (var property in data) {
                obj[property] = data[property];
            }
            data = obj;
        }
        let result = new ProjectInfo();
        result.init(data);
        return result;
    }

	override toJSON(data?: any) {
        data = typeof data === 'object' ? data : {};
        data["type"] = this.type ?? "ProjectInfo";
        data["north"] = this.north ?? 0;
        data["weather_urls"] = this.weatherUrls;
        data["location"] = this.location;
        data["ashrae_climate_zone"] = this.ashraeClimateZone;
        data["building_type"] = this.buildingType;
        data["vintage"] = this.vintage;
        data = super.toJSON(data);
        return instanceToPlain(data, { exposeUnsetFields: false });
    }

	async validate(): Promise<boolean> {
        const errors = await validate(this);
        if (errors.length > 0){
			const errorMessages = errors.map((error: TsValidationError) => Object.values(error.constraints || [error]).join(', ')).join('; ');
      		throw new Error(`Validation failed: ${errorMessages}`);
		}
        return true;
    }
}

File: src/CSharpSDK/Model/ProjectInfo.cs
/* 
 * HoneybeeSchema
 *
 * Contact: info@ladybug.tools
 */

//using System;
using System.Linq;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Runtime.Serialization;
using LBT.Newtonsoft.Json;
using LBT.Newtonsoft.Json.Converters;
using System.ComponentModel.DataAnnotations;

namespace HoneybeeSchema
{
    /// <summary>
    /// Project information.
    /// </summary>
    [Summary(@"Project information.")]
    [System.Serializable]
    [DataContract(Name = "ProjectInfo")] // Enables DataMember rules. For internal Serialization XML/JSON
    public partial class ProjectInfo : OpenAPIGenBaseModel, System.IEquatable<ProjectInfo>
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="ProjectInfo" /> class.
        /// </summary>
        [LBT.Newtonsoft.Json.JsonConstructorAttribute]
        // [System.Text.Json.Serialization.JsonConstructor] // for future switching to System.Text.Json
        protected ProjectInfo() 
        { 
            // Set readonly properties with defaultValue
            this.Type = "ProjectInfo";
        }
        /// <summary>
        /// Initializes a new instance of the <see cref="ProjectInfo" /> class.
        /// </summary>
        /// <param name="north">A number between -360 to 360 where positive values rotate the compass counterclockwise (towards the West) and negative values rotate the compass clockwise (towards the East).</param>
        /// <param name="weatherUrls">A list of URLs to zip files that includes EPW, DDY and STAT files. You can find these URLs from the EPWMAP. The first URL will be used as the primary weather file.</param>
        /// <param name="location">Project location. This value is usually generated from the information in the weather files.</param>
        /// <param name="ashraeClimateZone">Project location climate zone.</param>
        /// <param name="buildingType">A list of building types for the project. The first building type is considered the primary type for the project.</param>
        /// <param name="vintage">A list of building vintages (e.g. ASHRAE_2019, ASHRAE_2016).</param>
        public ProjectInfo
        (
            double north = 0D, List<string> weatherUrls = default, Location location = default, ClimateZones ashraeClimateZone = default, List<BuildingTypes> buildingType = default, List<EfficiencyStandards> vintage = default
        ) : base()
        {
            this.North = north;
            this.WeatherUrls = weatherUrls;
            this.Location = location;
            this.AshraeClimateZone = ashraeClimateZone;
            this.BuildingType = buildingType;
            this.Vintage = vintage;

            // Set readonly properties with defaultValue
            this.Type = "ProjectInfo";

            // check if object is valid, only check for inherited class
            if (this.GetType() == typeof(ProjectInfo))
                this.IsValid(throwException: true);
        }

	
	
        /// <summary>
        /// A number between -360 to 360 where positive values rotate the compass counterclockwise (towards the West) and negative values rotate the compass clockwise (towards the East).
        /// </summary>
        [Summary(@"A number between -360 to 360 where positive values rotate the compass counterclockwise (towards the West) and negative values rotate the compass clockwise (towards the East).")]
        // [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]  // For System.Text.Json  
        [Range(-360, 360)]
        [DataMember(Name = "north")] // For internal Serialization XML/JSON
        [JsonProperty("north", NullValueHandling = NullValueHandling.Ignore)] // For Newtonsoft.Json
        // [System.Text.Json.Serialization.JsonPropertyName("north")] // For System.Text.Json
        public double North { get; set; } = 0D;

        /// <summary>
        /// A list of URLs to zip files that includes EPW, DDY and STAT files. You can find these URLs from the EPWMAP. The first URL will be used as the primary weather file.
        /// </summary>
        [Summary(@"A list of URLs to zip files that includes EPW, DDY and STAT files. You can find these URLs from the EPWMAP. The first URL will be used as the primary weather file.")]
        // [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]  // For System.Text.Json  
        [DataMember(Name = "weather_urls")] // For internal Serialization XML/JSON
        [JsonProperty("weather_urls", NullValueHandling = NullValueHandling.Ignore)] // For Newtonsoft.Json
        // [System.Text.Json.Serialization.JsonPropertyName("weather_urls")] // For System.Text.Json
        public List<string> WeatherUrls { get; set; }

        /// <summary>
        /// Project location. This value is usually generated from the information in the weather files.
        /// </summary>
        [Summary(@"Project location. This value is usually generated from the information in the weather files.")]
        // [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]  // For System.Text.Json  
        [DataMember(Name = "location")] // For internal Serialization XML/JSON
        [JsonProperty("location", NullValueHandling = NullValueHandling.Ignore)] // For Newtonsoft.Json
        // [System.Text.Json.Serialization.JsonPropertyName("location")] // For System.Text.Json
        public Location Location { get; set; }

        /// <summary>
        /// Project location climate zone.
        /// </summary>
        [Summary(@"Project location climate zone.")]
        // [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]  // For System.Text.Json  
        [DataMember(Name = "ashrae_climate_zone")] // For internal Serialization XML/JSON
        [JsonProperty("ashrae_climate_zone", NullValueHandling = NullValueHandling.Ignore)] // For Newtonsoft.Json
        // [System.Text.Json.Serialization.JsonPropertyName("ashrae_climate_zone")] // For System.Text.Json
        public ClimateZones AshraeClimateZone { get; set; }

        /// <summary>
        /// A list of building types for the project. The first building type is considered the primary type for the project.
        /// </summary>
        [Summary(@"A list of building types for the project. The first building type is considered the primary type for the project.")]
        // [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]  // For System.Text.Json  
        [DataMember(Name = "building_type")] // For internal Serialization XML/JSON
        [JsonProperty("building_type", NullValueHandling = NullValueHandling.Ignore)] // For Newtonsoft.Json
        // [System.Text.Json.Serialization.JsonPropertyName("building_type")] // For System.Text.Json
        public List<BuildingTypes> BuildingType { get; set; }

        /// <summary>
        /// A list of building vintages (e.g. ASHRAE_2019, ASHRAE_2016).
        /// </summary>
        [Summary(@"A list of building vintages (e.g. ASHRAE_2019, ASHRAE_2016).")]
        // [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]  // For System.Text.Json  
        [DataMember(Name = "vintage")] // For internal Serialization XML/JSON
        [JsonProperty("vintage", NullValueHandling = NullValueHandling.Ignore)] // For Newtonsoft.Json
        // [System.Text.Json.Serialization.JsonPropertyName("vintage")] // For System.Text.Json
        public List<EfficiencyStandards> Vintage { get; set; }


        /// <summary>
        /// Returns the string presentation of the object
        /// </summary>
        /// <returns>String presentation of the object</returns>
        public override string ToString()
        {
            return "ProjectInfo";
        }


        /// <summary>
        /// Returns the string presentation of the object
        /// </summary>
        /// <returns>String presentation of the object</returns>
        public override string ToString(bool detailed)
        {
            if (!detailed)
                return this.ToString();
            
            var sb = new StringBuilder();
            sb.Append("ProjectInfo:\n");
            sb.Append("  Type: ").Append(this.Type).Append("\n");
            sb.Append("  North: ").Append(this.North).Append("\n");
            sb.Append("  WeatherUrls: ").Append(this.WeatherUrls).Append("\n");
            sb.Append("  Location: ").Append(this.Location).Append("\n");
            sb.Append("  AshraeClimateZone: ").Append(this.AshraeClimateZone).Append("\n");
            sb.Append("  BuildingType: ").Append(this.BuildingType).Append("\n");
            sb.Append("  Vintage: ").Append(this.Vintage).Append("\n");
            return sb.ToString();
        }


        /// <summary>
        /// Returns the object from JSON string
        /// </summary>
        /// <returns>ProjectInfo object</returns>
        public static ProjectInfo FromJson(string json)
        {
            var obj = JsonConvert.DeserializeObject<ProjectInfo>(json, JsonSetting.AnyOfConvertSetting);
            if (obj == null)
                return null;
            return obj.Type.ToLower() == obj.GetType().Name.ToLower() && obj.IsValid(throwException: true) ? obj : null;
        }




        /// <summary>
        /// Creates a new instance with the same properties.
        /// </summary>
        /// <returns>ProjectInfo object</returns>
        public virtual ProjectInfo DuplicateProjectInfo()
        {
            return FromJson(this.ToJson());
        }


        /// <summary>
        /// Creates a new instance with the same properties.
        /// </summary>
        /// <returns>OpenAPIGenBaseModel</returns>
        public override OpenAPIGenBaseModel DuplicateOpenAPIGenBaseModel()
        {
            return DuplicateProjectInfo();
        }


        /// <summary>
        /// Returns true if objects are equal
        /// </summary>
        /// <param name="input">Object to be compared</param>
        /// <returns>Boolean</returns>
        public override bool Equals(object input)
        {
            input = input is AnyOf anyOf ? anyOf.Obj : input;
            return this.Equals(input as ProjectInfo);
        }


        /// <summary>
        /// Returns true if ProjectInfo instances are equal
        /// </summary>
        /// <param name="input">Instance of ProjectInfo to be compared</param>
        /// <returns>Boolean</returns>
        public bool Equals(ProjectInfo input)
        {
            if (input == null)
                return false;
            return base.Equals(input) && 
                    Extension.Equals(this.North, input.North) && 
                    Extension.AllEquals(this.WeatherUrls, input.WeatherUrls) && 
                    Extension.Equals(this.Location, input.Location) && 
                    Extension.Equals(this.AshraeClimateZone, input.AshraeClimateZone) && 
                    Extension.AllEquals(this.BuildingType, input.BuildingType) && 
                    Extension.AllEquals(this.Vintage, input.Vintage);
        }


        /// <summary>
        /// Gets the hash code
        /// </summary>
        /// <returns>Hash code</returns>
        public override int GetHashCode()
        {
            unchecked // Overflow is fine, just wrap
            {
                int hashCode = base.GetHashCode();
                if (this.North != null)
                    hashCode = hashCode * 59 + this.North.GetHashCode();
                if (this.WeatherUrls != null)
                    hashCode = hashCode * 59 + this.WeatherUrls.GetHashCode();
                if (this.Location != null)
                    hashCode = hashCode * 59 + this.Location.GetHashCode();
                if (this.AshraeClimateZone != null)
                    hashCode = hashCode * 59 + this.AshraeClimateZone.GetHashCode();
                if (this.BuildingType != null)
                    hashCode = hashCode * 59 + this.BuildingType.GetHashCode();
                if (this.Vintage != null)
                    hashCode = hashCode * 59 + this.Vintage.GetHashCode();
                return hashCode;
            }
        }


    }
}
